torch/standalone/header_only/core/DeviceType.h

```diff
@@ -0,0 +1,288 @@
+  PrivateUse1 = 20, // PrivateUse1 device
+  // NB: If you add more devices:
+  //  - Change the implementations of DeviceTypeName and isValidDeviceType
+  //    in DeviceType.cpp
+  //  - Change the number below
+  COMPILE_TIME_MAX_DEVICE_TYPES = 21,
+};
+
+constexpr DeviceType kCPU = DeviceType::CPU;
+constexpr DeviceType kCUDA = DeviceType::CUDA;
+constexpr DeviceType kHIP = DeviceType::HIP;
+constexpr DeviceType kFPGA = DeviceType::FPGA;
+constexpr DeviceType kMAIA = DeviceType::MAIA;
+constexpr DeviceType kXLA = DeviceType::XLA;
+constexpr DeviceType kMPS = DeviceType::MPS;
+constexpr DeviceType kMeta = DeviceType::Meta;
+constexpr DeviceType kVulkan = DeviceType::Vulkan;
+constexpr DeviceType kMetal = DeviceType::Metal;
+constexpr DeviceType kXPU = DeviceType::XPU;
+constexpr DeviceType kHPU = DeviceType::HPU;
+constexpr DeviceType kVE = DeviceType::VE;
+constexpr DeviceType kLazy = DeviceType::Lazy;
+constexpr DeviceType kIPU = DeviceType::IPU;
+constexpr DeviceType kMTIA = DeviceType::MTIA;
+constexpr DeviceType kPrivateUse1 = DeviceType::PrivateUse1;
+
+// define explicit int constant
+constexpr int COMPILE_TIME_MAX_DEVICE_TYPES =
+    static_cast<int>(DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES);
+
+static_assert(
+    COMPILE_TIME_MAX_DEVICE_TYPES <= 21,
+    "Hey!  You seem to be adding a lot of new DeviceTypes.  The intent was "
+    "for this constant to reflect the actual number of DeviceTypes we support "
+    "in PyTorch; it's important that this number is not too large as we "
+    "use this to allocate stack arrays in some places in our code.  If you "
+    "are indeed just adding the 20th device type, feel free to change "
+    "the check to 32; but if you are adding some sort of extensible device "
+    "types registration, please be aware that you are affecting code that "
+    "this number is small.  Try auditing uses of this constant.");
+
+// We use both a mutex and an atomic here because:
+// (1) Mutex is needed during writing:
+//     We need to first check the value and potentially error,
+//     before setting the value (without any one else racing in the middle).
+//     It's also totally fine for this to be slow, since it happens exactly once
+//     at import time.
+// (2) Atomic is needed during reading:
+//     Whenever a user prints a privateuse1 device name, they need to read this
+//     variable. Although unlikely, we'll data race if someone else is trying to
+//     set this variable at the same time that another thread is print the
+//     device name. We could re-use the same mutex, but reading the atomic will
+//     be much faster.
+//
+// Note: Can not directly use inline to declar a global variable here, because
+//       it doesn't work for fbobjc build, error: declaration requires a global
+//       destructor C10_EXPORT is also needed to make sure the variables are
+//       visible to all files.
+inline C10_EXPORT std::atomic<bool>& privateuse1_backend_name_set() {
```

Comment:
To reviewers: most function bodies in this file are a straight-forward copy-paste from DeviceType.cpp, except the following three functions which are specially crafted to satisfy the global variable linking requirements on different platforms.

Reply:
What exactly are the global variable linking requirements on diff platforms and how do we make sure they are all still satisfied after this change?

Reply:
There were build-time errors on internal and OSS CIs. One of them is a fbobjc build error as I wrote in the comment. Another is a Windows linker error, "unresolved external symbol", which I solved by using C10_EXPORT.